### PR TITLE
spec out Ecotone L1 block attributes, l1 cost func, and gas price oracle changes

### PR DESCRIPTION
**Pull request by roberto-bayardo**
**Description**

Description

Addes specification of the L1 block attributes, l1 cost func, and gas price oracle following Ecotone upgrade

### CODE REVIEW

File: specs/exec-engine.md [hunk 1]
```diff
@@ -132,9 +137,43 @@ can be accessed in two interchangeable ways:
...
+ways:
+
+- read from the deposited L1 attributes (`l1BasefeeScalar`, `l1BlobBasefeeScalar`, `basefee`,
+  `blobBasefee`) of the current L2 block
+- read from the L1 Block Info contract (`0x4200000000000000000000000000000000000015`)
+  - using the respective solidity getter functions
+  - using direct storage-reads:
+    - basefee as big-endian `uint256` in slot `1`
+    - blobBasefee as big-endian `uint256` in slot `7`
+    - l1BasefeeScalar as big-endian `uint32` in first 4 bytes of slot `8`
```
Comment (tynes) [inline specs/exec-engine.md]: There is an inconsistency here - what does first 4 bytes mean? Relative to the msb or lsb? Big endian or little endian? `solc` prefers the "offset" language. The [solidity docs](https://docs.soliditylang.org/en/latest/internals/layout_in_storage.html) don't have any good examples. To get the offset, you can use the following command:

```
forge inspect SystemConfig storageLayout
```

Comment (tynes) [reply specs/exec-engine.md]: I don't think you need to specify msb or lsb or endianness if you use the same offset terminology that solc uses. iirc the offset is relative to the lsb

Comment (roberto-bayardo) [reply specs/exec-engine.md]: Yup, now that I finished the contract change I understand this better.  Updated to use slot+offset.

**Review by tynes — APPROVED** [review]
Looks good to me, just needs a linting fix

**Comment by 0xHansLee** [comment]
Hello @roberto-bayardo! I'm Hans from Kroma (https://github.com/kroma-network/kroma)

We are also preparing to apply 4844 based on your work. I would like to first say thank you since it has helped us a lot.

But I wonder how the L1 cost formula was derived. I thought that the `compressedTxSize` should be just a compressed transaction size, not multiplying 4 for zeros and 16 for ones since calldata is no longer used. Accordingly, I thought that the portion of l1Basefee is not needed, but only the blob fee part is needed for L1 cost (probably `compressedTxSize` * `l1BlobBasefee` * `l1BlobScalar`). Could you explain a little more about this?

**Comment by roberto-bayardo** [comment]
1. compressedTxSize is indeed an estimate of compressed transaction size: Note that the overall result is normalized by dividing by 16, though it's hard to see in the code as we do that division only at the very end to preserve precision under integer arithmetic.

2. basefee is still relevant as Blob transactions still use some amount of regular gas.  Also not all chains will use blob transactions -- especially for lower-traffic chains, blobs may not be cost effective since you pay the full price of a blob even if you don't need its entire capacity.

**Comment by 0xHansLee** [comment]
Thank you for your reply, @roberto-bayardo!

1. I see. Then, why the `compressedTxSize` is normalized by 16? I thought that it could be just `zeros + ones`.

2. I agree that basefee is still relevant even if the batch is posted as blob Tx. In case of using blob tx, however, shouldn't the cost related to basefee be unrelated to transaction size? Fees imposed based on compressed transaction size should be applied as blobfee. With [the formula](https://github.com/ethereum-optimism/optimism/pull/8737/files#diff-935783eb1ef1d5fbd583af5d07e028b6c2517bfcd5d84cafe7fcbc918d93b9b5R146) written in spec, fees are imposed on both basefee and blobfee based on the compressed transaction size, leading to a potential double charging.

If I'm misunderstanding anything, please let me know.

**Comment by roberto-bayardo** [comment]
1. Because non-zero bytes are weighted by 16.  Thus, if you have an entirely non-zero byte transaction (which we assume is highly incompressible), then uncompressed-size is identical to the compressed-size (that is, tx-len * 16 / 16 == tx-len).  If you have an entirely zero-byte transaction (which we assume is highly compressible) then uncompressed-size is estimated as tx-len / 4.  Everything else then is interpolated between .25 and 1.0 of the original transaction size based on the zero/non-zero ratio.  Obviously this is a very crude approximation of compressibility but it's the one Ethereum has implicitly used in its calldata cost function, as well as OP stack in the previous L1 data fee function. For the record we're not 100% settled on sticking with this scheme and I have done [some analysis on this scheme + better alternatives](https://github.com/roberto-bayardo/compression-analysis) we could adopt instead (though possibly in a later hardfork).

2. No, because the blob transactions aren't submitted one per L2 transaction, but 1 per full blob. Again it's helpful to think about this through extremes: if you can fit 1000 small transactions in a blob and only 1 very large L2 transaction in a blob, then it effectively costs 1000 times more in non-data-gas-fee to submit a single large L2 tx than it does 1 small tx.

**Comment by 0xHansLee** [comment]
@roberto-bayardo Yes, I understood that the blob tx for posting batch can include multiple L2 txs and it should cost proportional to transaction size.
 
I just thought that there are two parts in L1 cost imposed on L2 tx
1. blob fee proportional to transaction size
2. l1 basefee for regular gas of posting batch

But I may missed something. May I ask how the scalars (`l1BaseFeeScalar` and `l1BlobBasefeeScalar`) are determined?

**Pull request merged**